File: assignment1_task3_alice.m
%% Task 3 OFDM Receiver
% Based on: ProjectDescription_26_Part1.pdf and Guide Book Part I
% Receiver chain (from guide Fig. 7):
%   Demodulation -> LPF -> A/D -> Sync -> CP Removal -> FFT
%   -> Channel Estimation -> Equalization -> QPSK Decode -> Viterbi -> ASCII
clear; clc; close all;

%% Parameters (from assignment description)
fs   = 44100;    % sampling frequency [Hz]
fc   = 10000;    % carrier frequency [Hz]
Nsc  = 128;      % number of subcarriers
Ncp  = 20;       % cyclic prefix length [samples]
Tsym = 58e-3;    % OFDM symbol duration (data portion) [s]
Nofdm = Nsc + Ncp;

%% 1. Load signal
data = load('Signals_task3/signal3.mat');
r = double(data.R(:));

%% 2. Demodulation — down-convert from fc to baseband
n    = (0:length(r)-1)';
r_I  = 2 * r .* cos(2*pi*fc*n/fs);
r_Q  = -2 * r .* sin(2*pi*fc*n/fs);

%% 3. Low-pass filter (Butterworth, order 8, normalised cutoff 0.05)
[b, a] = butter(8, 0.05);
r_I_f = filtfilt(b, a, r_I);
r_Q_f = filtfilt(b, a, r_Q);

%% 4. Fractional downsampling — D = fs*Tsym/Nsc (not integer; use interpolation)
D     = fs * Tsym / Nsc;         % ≈ 19.9828
t_out = (0 : D : length(r_I_f)-1)';
r_bb  = interp1((0:length(r_I_f)-1)', r_I_f + 1j*r_Q_f, t_out, 'linear');
fprintf('D = %.4f,  length(r_bb) = %d,  ~%d OFDM symbols\n', ...
        D, length(r_bb), floor(length(r_bb)/Nofdm));

%% 5. Synchronization — periodic pilot correlation (Lecture 4)
% Pilot symbol has two identical halves of length Tper = Nsc/2 = 64
% mu(k) = |P(k)| / sqrt(E1(k)*E2(k)), start = argmax(mu)
Tper = Nsc / 2;
Nbb  = length(r_bb);
mu   = zeros(Nbb - 2*Tper, 1);
for k = 1:length(mu)
    s1 = r_bb(k         : k+Tper-1);
    s2 = r_bb(k+Tper    : k+2*Tper-1);
    gamma = sum(s1 .* conj(s2));
    E1    = sum(abs(s1).^2);
    E2    = sum(abs(s2).^2);
    mu(k) = abs(gamma) / sqrt(E1*E2 + 1e-10);
end
[~, T0] = max(mu);          % T0: 1-indexed sample where sync metric peaks
t0 = T0 - Ncp - 1;          % back up by CP to reach start of pilot symbol
fprintf('Sync peak T0 = %d,  frame start t0 = %d\n', T0, t0);

%% 6. Frame extraction and CP removal
n_sym  = floor((Nbb - t0 + 1) / Nofdm);
frame  = r_bb(t0 : t0 + n_sym*Nofdm - 1);
blocks = reshape(frame, Nofdm, n_sym)';   % n_sym x Nofdm
ofdm_no_cp = blocks(:, Ncp+1:end);        % n_sym x Nsc  (CP removed)
fprintf('Total OFDM symbols: %d\n', n_sym);
% Layout: row 1 = pilot, row 2 = length, rows 3..end = data

%% 7. FFT — frequency-domain representation
X = fft(ofdm_no_cp, Nsc, 2);
% X(1,:) = pilot symbol
% X(2,:) = length symbol
% X(3:end,:) = data symbols

%% 8. Reconstruct pilot sequence (identical to transmitter)
% Transmitter: randn('state',100); P = sign(randn(1,Nsc/2))
randn('state', 100);
P = sign(randn(1, Nsc/2));   % 64-element sequence of +1/-1

%% 9. Channel estimation
% Pilot symbol: x_pilot(1:2:Nsc) = 2*P  (odd MATLAB indices carry pilots, even = 0)
%   => X(1, 2k-1) = H(2k-1) * 2 * P(k)  for k = 1..64
pilot_sc = 1:2:Nsc;                        % MATLAB 1-indexed: 1,3,5,...,127
H_pil    = X(1, pilot_sc) ./ (2 * P);     % H at all 64 pilot subcarriers

% Length symbol: subcarriers 16..128 carry known 1+j
%   (10-bit length → conv encode → zero-pad to 2*Nsc bits → QPSK; first 15 sc carry
%    length bits, remaining 113 carry zero-bits which QPSK-map to 1+j)
len_sc = 16:Nsc;
H_len  = X(2, len_sc) ./ (1 + 1j);       % H at subcarriers 16..128 (113 values)

% Combine into full channel estimate for all Nsc subcarriers
H_full = zeros(1, Nsc);
for k = 1:Nsc
    is_pil = (mod(k, 2) == 1);   % odd MATLAB index → pilot measurement available
    is_len = (k >= 16);           % in length-symbol known region
    if is_pil && is_len
        % Both estimates available — average them
        H_full(k) = 0.5 * (H_pil((k+1)/2) + H_len(k-15));
    elseif is_pil
        H_full(k) = H_pil((k+1)/2);
    elseif is_len
        H_full(k) = H_len(k-15);
    % else: even subcarrier below 16 — filled by interpolation below
    end
end
% Interpolate even subcarriers 2,4,...,14 from their odd neighbours
for k = 2:2:14
    H_full(k) = 0.5 * (H_full(k-1) + H_full(k+1));
end
% Smooth real and imaginary parts to reduce per-subcarrier noise
H_full = sgolayfilt(real(H_full), 3, 7) + 1j * sgolayfilt(imag(H_full), 3, 7);

%% 10. Equalization (divide by channel estimate)
X_len_eq  = X(2,:) ./ H_full;         % length symbol equalized  (1 x Nsc)
X_data_eq = X(3:end,:) ./ H_full;     % data symbols equalized   (n_data x Nsc)

%% 11. QPSK demodulation
% Mapping: 00->1+j, 10->-1+j, 11->-1-j, 01->1-j
% First bit = sign(real):  0 if real>0,  1 if real<0
% Second bit = sign(imag): 0 if imag>0,  1 if imag<0
% Bit stream is serialised symbol-by-symbol, subcarrier-by-subcarrier within each symbol

% Length symbol bits (1 x Nsc -> 1 x 2*Nsc bit vector)
bits_len = zeros(1, Nsc*2);
bits_len(1:2:end) = double(real(X_len_eq) < 0);
bits_len(2:2:end) = double(imag(X_len_eq) < 0);

% Data symbol bits — row-major flatten (symbol 0 all scs, then symbol 1, ...)
% In MATLAB (column-major): transpose first, then column-flatten = row-major of original
X_data_flat = reshape(X_data_eq.', 1, []);
bits_data = zeros(1, length(X_data_flat)*2);
bits_data(1:2:end) = double(real(X_data_flat) < 0);
bits_data(2:2:end) = double(imag(X_data_flat) < 0);

%% 12. Viterbi decoding — convolutional code (77,45) octal, rate 1/2, K=6
% Encoder terminates with K-1=5 zero bits, so use 'term' mode
trellis = poly2trellis(6, [77 45]);

% --- Decode message length ---
% Length: 10 data bits + 5 tail bits -> 15 input bits -> 30 encoded bits
dec_len = vitdec(bits_len(1:30), trellis, 35, 'term', 'hard');
l_m = sum(dec_len(1:10)' .* (2.^(9:-1:0))');
fprintf('Decoded message length: %d characters\n', l_m);

% --- Decode message (search around decoded length for best result) ---
best_msg = ''; best_score = -1; best_lm = -1;
for lm_try = max(1, l_m-30) : min(400, l_m+30)
    needed = 14*lm_try + 10;   % l_m*7 data bits + 5 tail bits, encoded at rate 1/2
    if needed > length(bits_data), continue; end
    dec = vitdec(bits_data(1:needed), trellis, 35, 'term', 'hard');
    msg_bits = dec(1 : 7*lm_try);   % vitdec 'term' output = needed/2 - (K-1) = 7*lm_try
    chars = zeros(1, lm_try);
    for i = 1:lm_try
        b7 = msg_bits((i-1)*7+1 : i*7);
        chars(i) = sum(b7(:)' .* (2.^(6:-1:0)));
    end
    msg   = char(chars);
    score = sum(isletter(msg) | msg == ' ') / max(1, lm_try);
    if score > best_score || (score == best_score && lm_try > best_lm)
        best_score = score;
        best_msg   = msg;
        best_lm    = lm_try;
    end
end

fprintf('\nBest decode  (length=%d, alpha-rate=%.3f):\n%s\n', best_lm, best_score, best_msg);

File: ipynb_to_py/assignment1_task3_alice.py
import numpy as np
import matplotlib.pyplot as plt
import scipy.io
from numpy.lib.stride_tricks import sliding_window_view

# Load signal
mat_path = 'Signals_task3/signal3.mat'
data = scipy.io.loadmat(mat_path)
r = data['R'].flatten()

# System parameters
fs = 44100
fc = 10000
Nsc = 128
Ncp = 20
Tsym = 58 * 10**-3

# **Demodulation**
n = np.arange(len(r))
r_I =  2 * r * np.cos(2 * np.pi * fc * n / fs)
r_Q = -2 * r * np.sin(2 * np.pi * fc * n / fs)
from scipy.signal import butter, freqz, filtfilt

b, a = butter(8, 0.05)
r_I_filt = filtfilt(b, a, r_I)
r_Q_filt = filtfilt(b, a, r_Q)

w, h = freqz(b, a)
plt.plot(w, 20 * np.log10(abs(h)))
plt.title('Butterworth Filter Frequency Response')
plt.xlabel('Normalized Frequency (xπ rad/sample)')
plt.ylabel('Amplitude (dB)')
plt.grid()
plt.show()

# **A/D Conversion and Down Sampling**
D = fs * Tsym / Nsc   # exact ratio, not rounded
idx = np.arange(0, len(r_I_filt), D)
r_I_down = np.interp(idx, np.arange(len(r_I_filt)), r_I_filt)
r_Q_down = np.interp(idx, np.arange(len(r_Q_filt)), r_Q_filt)
r_bb = r_I_down + 1j * r_Q_down
print(f'D={D:.4f}, r_bb length={len(r_bb)}, ~{len(r_bb)//(Nsc+Ncp)} OFDM symbols')

# **Synchronization and Removal of the Cyclic Prefix**
Tper = Nsc // 2  # = 64, pilot symbol period

wins  = sliding_window_view(r_bb, 2 * Tper)
gamma = np.sum(wins[:, :Tper] * np.conj(wins[:, Tper:]), axis=1)
phi1  = np.sum(np.abs(wins[:, :Tper])**2, axis=1)
phi2  = np.sum(np.abs(wins[:, Tper:])**2, axis=1)
mu    = np.abs(gamma) / np.sqrt(phi1 * phi2 + 1e-10)

T0 = np.argmax(mu)          # argmax falls at start of pilot data (after CP)
t0 = T0 - Ncp - 1           # start of pilot CP; -1 corrects a sub-sample sync bias
print(f'Pilot starts at: {t0}, OFDM data at: {T0}')

Nofdm   = Nsc + Ncp
n_sym   = (len(r_bb) - t0) // Nofdm
ofdm_blocks = r_bb[t0 : t0 + n_sym*Nofdm].reshape(n_sym, Nofdm)
ofdm_data   = ofdm_blocks[:, Ncp:]
print(f'Total OFDM symbols: {n_sym}')

# **FFT**
X = np.fft.fft(ofdm_data, axis=1)

# **Channel estimation**
from scipy.signal import savgol_filter

# H from zero-padded subcarriers 15-127 of the length OFDM symbol (X[1])
# Those subcarriers carry all-zero bits → QPSK symbol = 1+j
H_from_len = X[1, 15:] / (1+1j)   # shape (113,)
len_pos = np.arange(15, 128)

# Interpolate real and imaginary parts separately to all 128 subcarriers
# (more stable than interpolating magnitude and phase)
all_pos = np.arange(Nsc)
H_init = (np.interp(all_pos, len_pos, H_from_len.real)
        + 1j * np.interp(all_pos, len_pos, H_from_len.imag))

# Use ALL 64 even subcarriers from the pilot symbol to improve H estimate
# Pilot: X[0, 2k] = 2 * P[k] * H[2k], where P[k] ∈ {+1, -1}
even_idx = np.arange(0, Nsc, 2)
raw = X[0, even_idx]                       # received pilot at even subcarriers
ratio = raw / (2 * H_init[even_idx])
P = np.where(np.abs(ratio.real) >= np.abs(ratio.imag),
             np.sign(ratio.real), np.sign(ratio.imag))
P = np.where(P == 0, 1.0, P)
H_from_pilot = raw / (2 * P)              # H estimate at all 64 even subcarriers

# Combine estimates: average where both are available (even sc >= 15)
known_set = set(len_pos.tolist())
H_full = np.empty(Nsc, dtype=complex)
for k in range(Nsc):
    in_len = k in known_set
    in_pil = (k % 2 == 0)
    if in_len and in_pil:
        H_full[k] = 0.5 * (H_from_len[k - 15] + H_from_pilot[k // 2])
    elif in_len:
        H_full[k] = H_from_len[k - 15]
    elif in_pil:
        H_full[k] = H_from_pilot[k // 2]
    else:
        H_full[k] = 0  # odd subcarrier below 15 — filled next

# Fill odd subcarriers 1,3,...,13 by averaging their even neighbours
for k in range(1, 15, 2):
    H_full[k] = 0.5 * (H_full[k - 1] + H_full[k + 1])

# Smooth both parts to reduce per-subcarrier noise
H_full = savgol_filter(H_full.real, 7, 3) + 1j * savgol_filter(H_full.imag, 7, 3)

plt.scatter(X[2:8].real.flatten(), X[2:8].imag.flatten(), s=1)
plt.title('Raw data symbols (before equalization)'); plt.xlabel('Real'); plt.ylabel('Imaginary')
plt.grid(); plt.show()

# **Remove pilots**
X_data = X[1:]   # length symbol + all remaining symbols
plt.plot(np.abs(X[2:8]).T)
plt.title('Raw magnitude of data OFDM symbols'); plt.xlabel('Subcarrier'); plt.grid(); plt.show()

# **Equalization**
X_eq = X_data / H_full   # equalize: divide each row by H estimate

# Plot only the 6 actual data symbols (X_eq[0]=length, X_eq[1:7]=data)
plt.scatter(X_eq[1:7].real.flatten(), X_eq[1:7].imag.flatten(), s=5)
plt.xlim(-3, 3); plt.ylim(-3, 3)
plt.axhline(0, color='k', lw=0.5); plt.axvline(0, color='k', lw=0.5)
plt.title('Equalized constellation (6 data symbols only)')
plt.grid(); plt.show()

bits = []
for sym in X_eq.flatten():
    bits.append('1' if sym.real < 0 else '0')
    bits.append('1' if sym.imag < 0 else '0')

# **QPSK DEMOD**
bits = []
for sym in X_eq.flatten():
    bits.append('1' if sym.real < 0 else '0')
    bits.append('1' if sym.imag < 0 else '0')

# **Decoding and ASCII**
from commpy.channelcoding import Trellis, viterbi_decode

bits_arr  = np.array(bits).astype(int)
bits_len  = bits_arr[:256].astype(float)
bits_data = bits_arr[256:].astype(float)

best_msg, best_score, best_l_m = None, -1, None

for g2 in (0o45, 0o51):
    trellis = Trellis(memory=np.array([5]), g_matrix=np.array([[0o77, g2]]))

    # Decode the message length from the length OFDM symbol
    decoded_len = viterbi_decode(bits_len, trellis, decoding_type='hard')
    l_m_auto = int(''.join(decoded_len[:10].astype(int).astype(str)), 2)
    print(f'g2=0o{g2:o}: auto-detected length = {l_m_auto}')

    # Search around the decoded length in case of length-symbol decode errors
    for l_m in set([l_m_auto] + list(range(max(1, l_m_auto - 20), l_m_auto + 21))):
        if l_m < 1 or l_m > 400:
            continue
        needed = 14 * l_m + 10
        if needed > len(bits_data):
            continue
        # Use tb_depth=35 (larger than default 25) for more reliable traceback
        decoded_msg = viterbi_decode(bits_data[:needed], trellis, tb_depth=35, decoding_type='hard')
        msg_bits = decoded_msg[:l_m * 7].astype(int).astype(str)
        message = ''.join(
            chr(int(''.join(msg_bits[i:i+7]), 2))
            for i in range(0, l_m * 7, 7)
            if len(msg_bits[i:i+7]) == 7
        )
        score = sum(c.isalpha() or c == ' ' for c in message) / max(1, len(message))
        if score > best_score:
            best_score, best_msg, best_l_m = score, message, l_m

print(f'\nBest decode (length={best_l_m}, score={best_score:.3f}):')
print(best_msg)
